lib/render/ParticleRenderer.cpp

```diff
@@ -426,21 +375,41 @@ void ParticleRenderer::_generateParticleData(const Grid* grid, const std::vector
             continue;
         }
 
+        _streamSizes.push_back(4);
+
         const float value = grid->GetValueAtIndex(*node);
         grid->GetUserCoordinates(*node, coordsBuf);
-        const glm::vec3 p(coordsBuf[0], coordsBuf[1], coordsBuf[2]);
-
-        _particles.push_back({glm::vec3(p[0], p[1], p[2]), value});
+        const glm::vec3 p1(coordsBuf[0], coordsBuf[1], coordsBuf[2]);
 
         if (showDir) {
             const glm::vec3 n(*(dirs[0]), *(dirs[1]), *(dirs[2]));
-            const glm::vec3 p2 = p + n * dirScale;
+            const glm::vec3 p2 = p1 + n * dirScale;
+            glm::vec3 prep(-normalize(p1 - p2) + p2);
+            glm::vec3 post( normalize(p1 - p2) + p1);
+
+            _particles.push_back({prep, value});
             _particles.push_back({glm::vec3(p2[0], p2[1], p2[2]), value});
+            _particles.push_back({glm::vec3(p1[0], p1[1], p1[2]), value});
+            _particles.push_back({post, value});
 
             for (auto &it : dirs) ++it;
         }
-        else _particles.push_back({glm::vec3(p[0], p[1], p[2]), value});
+        else {
+            glm::vec3 prep(-normalize(p1 - p1) + p1);
+            _particles.push_back({prep, value});
+            _particles.push_back({glm::vec3(p1[0], p1[1], p1[2]), value});
+            _particles.push_back({glm::vec3(p1[0], p1[1], p1[2]), value});
+            _particles.push_back({prep, value});
+        }
     }
+    
+    assert(glIsVertexArray(_VAO) == GL_TRUE);
+    assert(glIsBuffer(_VBO) == GL_TRUE);
+
+    glBindVertexArray(_VAO);
+    glBindBuffer(GL_ARRAY_BUFFER, _VBO);
+    glBufferData(GL_ARRAY_BUFFER, sizeof(_vertex) * _particles.size(), _particles.data(), GL_STREAM_DRAW);
```

Comment:
The struct `_vertex` follows the naming convention of a private member variable which was confusing when reviewing.

Reply:
Hmm.  I thought we put underscores on all private members, even definitions.  I've capitalized it to be _Vertex_ instead, which seems like what is being done in the Volume renderer, if we'd like to make that our policy.